### PR TITLE
PortScannerDetector

### PR DESCRIPTION
**Pull request by moltzaum**
* Modularized functions into their own files
* Using default python dictionary instead of computing index manually
* Using channels among threads to pass data in a threadsafe way
* Removed timer thread (put into detector thread function)
* Added dissector thread so that the sniffer won't drop packets while dissecting
* Tested running of program across a span of 5 minutes for a -T4 speed on nmap
* Made [Ctrl+C] exit the program smoothly (as opposed to needing to Ctrl+C for each thread)

**EDIT: Fixed this, see below**
At this point, the dst_ip is not used. It is not clear if it could be used.
This is something to test across at least 3 machines on a network.

**EDIT: Tested this, see below**
I think different speeds should be tested, and it would be interesting to
even test multiple port scanners at once. At the moment, the port scanner
still needs to be updated to scan at different rates, since in the end it
seems we should be using that over nmap.

Note: I also removed comments temporarily. This is why I haven't deleted
the original file yet. They can be added back, but I didn't want to since
I would have felt compelled to make similar comments for new functions.
TL;DR The result of this is a combination of OCD and Laziness

### CODE REVIEW

File: detector/sniffer.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+# Note: AF_PACKET does not work on Mac, but works on Linux
+_packets = socket.socket(socket.AF_PACKET, socket.SOCK_RAW, socket.htons(0x0800))
+
+# This function sniffs data and immediately puts it on a queue to be processed. It does this
+# so that it doesn't miss other incoming packets while processing the data or while waiting
+# to receive the table from another channel.
+# 
+# Note: not all packets will be received, but it allows us to capture x2 the number of
+# packets from before. It may also be beneficial to start two sniffers at once. When I tested
+# that, I got an extra 100 packets. Is the recvfrom method atomic as to prevent duplicates?
```
Comment (moltzaum) [inline detector/sniffer.py]: In the end, I think the function was improved further so that we capture every packet. Now I see 1000 packets when I run the nmap scan. nmap may only send a connection to this many ports. With our own port scanner, we can test the exact number and limit.

File: detector/sniffer.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+            elif ip_protocol == IPProtocol.UDP:
+                src_port, dst_port = udp_dissect(transport_data)
+                # do nothing..
+                
+                # After we implement a scanner detector for TCP,
+                # we may want to add UDP on top of that. We should
+                # distinguish TCP and UDP ports in the table we
+                # have, and we might need to incorporate logic
+                # taken from ICMP? (I don't think think this is
+                # the case)
```
Comment (moltzaum) [inline detector/sniffer.py:62]: Note: I don't actually think we need to implement a UDP port scanner.

**Comment by moltzaum** [comment]
I tested this with all the nmap speeds just now. -T3 gets detected pretty quickly, and -T2 gets detected, but it takes awhile. -T1 does not get detected with the current rates set. In every case, the fan out rate that trips first is when the total count reaches above 300. Surprisingly, this is also true for -T5.

**Comment by moltzaum** [comment]
I just spun up two kali vms. The portscanner can only detect ports being scanned on the current machine, so it seems we really can disregard the dst_ip.

EDIT:
From the assignment description: "This tool will run on Kali VM and detect every host that is performing port scanning in the LAN where the Kali VM is located in."

From the assignment description: "Note: You may not be able to see if your own machine is the port-scanner. Try to see if you can make this work. I will give additional points for it."

Ok, maybe there is some more stuff we can try out. I think this might work? https://stackoverflow.com/q/6067405/9990099

EDIT2: I figured out how to set the scanner into promiscuous mode. This allows us to detect all packets on the network, including the host. I'm just not sure at the moment if it can detect packets on the host fast enough.

File: detector/sniffer.py (hunk 1 above)
Comment (moltzaum) [reply detector/sniffer.py]: Tested again with my portscanner from lab 2. All the ports I scanned were captured. I think its odd that nmap only scans for 1000 and not 1024.

**Comment by moltzaum** [comment]
From the assignment: "Note: You may not be able to see if your own machine is the port-scanner. Try to see if you can make this work. I will give additional points for it."

Ok, maybe there is some more stuff we can try out. I think this might work? https://stackoverflow.com/q/6067405/9990099

**Comment by moltzaum** [comment]
I found a new limitation with promiscuous mode on. For some reason the program marks both the attacker and victim as a port scanner. As far as I can tell right now, the victim sends a response for every packet sent its way. Maybe when we dissect the ethernet packet there will be information on whether or not a packet is a response. I know this is the case for ICMP, but I'm not sure about TCP. EDIT: We need to filter out packets that don't have the `SYN` packet set. EDIT2: Fixed

File: detector/sniffer.py (hunk 1 above)
Comment (moltzaum) [reply detector/sniffer.py]: After some other tests, sometimes not all ports are picked up and I think it is due to factors outside of the program. I was running the wireshark viewer in the background and only got 700 packets in.
